### PR TITLE
H-1146: Require a permission to instantiate an entity per entity type

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Certain types must not be created without a specific flow, otherwise the database would result in an inconsistent state. The most important example here is the `User` entity. 

## 🔍 What does this change?

- Implement a relationship and permission to allow the instantiation of a type
- Add a non-optional parameter to the `createEntityType` and `updateEntityType` methods where a list of subjects can be specified. These will be allowed to instantiate the type
- Call the function with `[{ kind: "public" }]` everywhere, that means that everyone is allowed to instantiate an entity of the specific type (it still requires permission to create an entity in the web).
- **This PR does not special case any system types such as User***

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

### CODE REVIEW

File: apps/hash-api/src/graph/ontology/primitive/entity-type.ts
Comment (benwerner01) [inline apps/hash-api/src/graph/ontology/primitive/entity-type.ts:54]: Is there a way we can avoid casting here? This could lead to us forgetting to update something in the Node API if the return type changes.

Comment (TimDiekmann) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:54]: This was raised before when I implemented this for entities [a while back](https://github.com/hashintel/hash/pull/3403/files). We had a quite lengthy discussion about this in standup and it's not possible to avoid the casts as we need to cast from `string` to `AccountId`. The whole point of the cast is to make it more safe.

In this PR I only moved the function to the top so I can use it later.

File: apps/hash-api/src/graph/ontology/primitive/entity-type.ts
Comment (TimDiekmann) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:96]: I feel pretty strongly to not use a default value here as this is a permission specific argument. (Generally, I got very used to not have default arguments at all). With this being a default argument it's easy to just make an entity type public while this is the most conservative way to drive the API. It's very easy to miss default parameters if you just copy the method call and it's even very likely that you forget that this is a thing. If we'd have a default argument I'd argue for `[]`, but then we would need to pass a parameter almost in every case (there is not really a use-case for `[]` I think).

File: apps/hash-api/src/graph/ontology/primitive/entity-type.ts
Comment (TimDiekmann) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:251]: Instantiators are tied to the versioned URL, not to the base URL, so it's required to specify it. On default arguments the reasoning from above applies as well.

File: apps/hash-api/src/graph/ontology/primitive/entity-type.ts
Comment (TimDiekmann) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:96]: Also worth mentioning that this is only required in a very few places.

File: apps/hash-api/src/graph/ontology/primitive/entity-type.ts
Comment (CiaranMn) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:251]: Why tied to the versioned URL, rather than the base URL? The base URL would seem more consistent with entities, where permissions are associated with the `baseId` (the `EntityId`) rather than an edition.

Once we introduce the ability to share editing permissions for a type with a specific user, we'll want this to apply to the type and future versions – otherwise every time a type is updated, we'll have to retrieve and apply all existing permissions on it to avoid changing them unintentionally.

Comment (TimDiekmann) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:251]: We currently use `VersionedUrl` everywhere for ontology type permission. I can change this to be the `BaseUrl` but that's a bigger change.

Comment (TimDiekmann) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:256]: Binding it to the `BaseUrl` implies that it's not possible to for example restrict older versions of a type to be instantiated.

File: apps/hash-api/src/graph/ontology/primitive/entity-type.ts
Comment (benwerner01) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:98]: > Also worth mentioning that this is only required in a very few places.

Well it'll be required wherever we create entity types, unless we set a default on a higher abstraction level which I suppose is what we are essentially doing (we are setting the default in the GraphQL resolvers in this case). I don't feel very strongly about this, so happy to leave as is.

File: apps/hash-api/src/graph/ontology/primitive/entity-type.ts
Comment (TimDiekmann) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:54]: We can change that but I prefer to not doing this as a drive-by in this PR as this PR is unrelated.

Comment (benwerner01) [reply apps/hash-api/src/graph/ontology/primitive/entity-type.ts:54]: Okay let's create a follow-up then, I'm keen to avoid adding more casts where possible given the difficulties we had finding and removing the various `Subgraph` casts

**Review by benwerner01 — DISMISSED** [review]
Thanks Tim, the TS code makes sense to me. Maybe worth waiting for @CiaranMn to get final approval as there seems to be an unresolved conversation we may want to capture in linear.

**Review by CiaranMn — APPROVED** [review]

**Pull request merged**
